### PR TITLE
Adds support for checksums in streamed request trailers

### LINKED ISSUE #954: Multipart Uploads using Checksums
**Describe the bug**

I'm using the [s3fs library](https://github.com/fsspec/s3fs) that makes use of aiobotocore underneath. I'm running the library like this:

```
fs = s3fs.S3FileSystem(s3_additional_kwargs={'ChecksumAlgorithm': 'SHA256'})
with fs.open(f'some_bucket/some_object', 'w', block_size=10 * 1024 * 1024) as f:
  f.write("hello")
```

Not supplying the kwargs works fine, yet I'm expecting it to work with the latest ChecksumAlgorithm feature enabled.

I'm receiving several errors from aiobotocore:
```
Traceback (most recent call last):
  File "Python39\lib\site-packages\aiohttp\client_reqrep.py", line 510, in update_body_from_data
    body = payload.PAYLOAD_REGISTRY.get(body, disposition=None)
  File "Python39\lib\site-packages\aiohttp\payload.py", line 118, in get
    raise LookupError()
aiohttp.payload.LookupError


Traceback (most recent call last):
  File "Python39\lib\site-packages\aiobotocore\httpsession.py", line 178, in send
    response = await self._session.request(
  File "Python39\lib\site-packages\aiohttp\client.py", line 507, in _request
    req = self._request_class(
  File "Python39\lib\site-packages\aiohttp\client_reqrep.py", line 313, in __init__
    self.update_body_from_data(data)
  File "Python39\lib\site-packages\aiohttp\client_reqrep.py", line 512, in update_body_from_data
    body = FormData(body)()
  File "Python39\lib\site-packages\aiohttp\formdata.py", line 37, in __init__
    self.add_fields(*fields)
  File "Python39\lib\site-packages\aiohttp\formdata.py", line 107, in add_fields
    raise TypeError(
TypeError: Only io.IOBase, multidict and (name, file) pairs allowed, use .add_field() for passing more complex parameters, got <botocore.httpchecksum.AwsChunkedWrapper object at 0x000001187CDEC0A0>


Traceback (most recent call last):  
  < my code snipped >
    f.write("hello")
  File "Python39\lib\site-packages\fsspec\spec.py", line 1664, in close
    self.flush(force=True)
  File "Python39\lib\site-packages\fsspec\spec.py", line 1535, in flush
    if self._upload_chunk(final=force) is not False:
  File "Python39\lib\site-packages\s3fs\core.py", line 2095, in _upload_chunk
    self.commit()
  File "Python39\lib\site-packages\s3fs\core.py", line 2110, in commit
    write_result = self._call_s3(
  File "Python39\lib\site-packages\s3fs\core.py", line 1962, in _call_s3
    return self.fs.call_s3(method, self.s3_additional_kwargs, *kwarglist, **kwargs)
  File "Python39\lib\site-packages\fsspec\asyn.py", line 86, in wrapper
    return sync(self.loop, func, *args, **kwargs)
  File "Python39\lib\site-packages\fsspec\asyn.py", line 66, in sync
    raise return_result
  File "Python39\lib\site-packages\fsspec\asyn.py", line 26, in _runner
    result[0] = await coro
  File "Python39\lib\site-packages\s3fs\core.py", line 332, in _call_s3
    return await _error_wrapper(
  File "Python39\lib\site-packages\s3fs\core.py", line 137, in _error_wrapper
    raise err
  File "Python39\lib\site-packages\s3fs\core.py", line 110, in _error_wrapper
    return await func(*args, **kwargs)
  File "Python39\lib\site-packages\aiobotocore\client.py", line 251, in _make_api_call
    http, parsed_response = await self._make_request(
  File "Python39\lib\site-packages\aiobotocore\client.py", line 271, in _make_request
    return await self._endpoint.make_request(operation_model, request_dict)
  File "Python39\lib\site-packages\aiobotocore\endpoint.py", line 80, in _send_request
    while await self._needs_retry(attempts, operation_model,
  File "Python39\lib\site-packages\aiobotocore\endpoint.py", line 213, in _needs_retry
    responses = await self._event_emitter.emit(
  File "Python39\lib\site-packages\aiobotocore\hooks.py", line 50, in _emit
    response = await resolve_awaitable(handler(**kwargs))
  File "Python39\lib\site-packages\aiobotocore\_helpers.py", line 11, in resolve_awaitable
    return await obj
  File "Python39\lib\site-packages\aiobotocore\retryhandler.py", line 89, in _call
    if await resolve_awaitable(self._checker(**checker_kwargs)):
  File "Python39\lib\site-packages\aiobotocore\_helpers.py", line 11, in resolve_awaitable
    return await obj
  File "Python39\lib\site-packages\aiobotocore\retryhandler.py", line 107, in _call
    should_retry = await self._should_retry(attempt_number, response,
  File "Python39\lib\site-packages\aiobotocore\retryhandler.py", line 129, in _should_retry
    return await resolve_awaitable(
  File "Python39\lib\site-packages\aiobotocore\_helpers.py", line 11, in resolve_awaitable
    return await obj
  File "Python39\lib\site-packages\aiobotocore\retryhandler.py", line 146, in _call
    checker(attempt_number, response,
  File "Python39\lib\site-packages\botocore\retryhandler.py", line 247, in __call__
    return self._check_caught_exception(
  File "Python39\lib\site-packages\botocore\retryhandler.py", line 416, in _check_caught_exception
    raise caught_exception
  File "Python39\lib\site-packages\aiobotocore\endpoint.py", line 149, in _do_get_response
    http_response = await self._send(request)
  File "Python39\lib\site-packages\aiobotocore\endpoint.py", line 229, in _send
    return await self.http_session.send(request)
  File "Python39\lib\site-packages\aiobotocore\httpsession.py", line 220, in send
    raise HTTPClientError(error=e)
botocore.exceptions.HTTPClientError: An HTTP Client raised an unhandled exception: Only io.IOBase, multidict and (name, file) pairs allowed, use .add_field() for passing more complex parameters, got <botocore.httpchecksum.AwsChunkedWrapper object at 0x00000197488C50A0>

Process finished with exit code 1

```

**Checklist**
- [x] I have reproduced in environment where `pip check` passes without errors
- [x] I have provided `pip freeze` results
- [x] I have provided sample code or detailed way to reproduce
- [ ] I have tried the same code in botocore to ensure this is an aiobotocore specific issue
- [ ] I have tried similar code in aiohttp to ensure this is is an aiobotocore specific issue
- [ ] I have checked the latest and older versions of aiobotocore/aiohttp/python to see if this is a regression / injection

**pip freeze results**

```
aiobotocore==2.3.4
aiohttp==3.8.1
aioitertools==0.10.0
aiosignal==1.2.0
async-timeout==4.0.2
attrs==21.4.0
azure-core==1.24.2
azure-storage-blob==12.13.0
black==22.6.0
boto3==1.24.27
botocore==1.27.27
certifi==2022.6.15
cffi==1.15.1
charset-normalizer==2.1.0
click==8.1.3
colorama==0.4.5
cryptography==37.0.4
DateTime==4.5
frozenlist==1.3.0
fsspec==2022.5.0
idna==3.3
isodate==0.6.1
```

**Environment:**
 - Python Version: 3.9
 - OS name and version: win11 

**Additional context**

### PR DESCRIPTION
**Pull request by terrycain**
### Description of Change
Adds support for S3 streaming uploads to support checksums in appended on the end of the request stream (trailers)

Running a snippet like the following will use streaming uploads and also the new SHA256 checksums
```python
async def main():
    session = get_session()
    async with session.create_client('s3', region_name='eu-west-2') as s3_client:
        data = b'test1234'
        digest = hashlib.sha256(data).digest()

        resp = await s3_client.put_object(Bucket="BUCKET", Key='foobarbaz', Body=data, ChecksumAlgorithm='SHA256')
        sha256_trailer_checksum = base64.b64decode(resp['ChecksumSHA256'])

        assert digest == sha256_trailer_checksum
```
Note - when using Moto or an S3 endpoint not using HTTPS, the checksum will be added to the header, not the trailers as the headers are signed, the trailer is not (hence the requirements for HTTPS)

The problem comes with `botocore.httpchecksum.AwsChunkedWrapper` ending up as the body passed to `aiohttp` and it not liking it. We now subclass that wrapper class and add an async iterator implementation. 

Fixes #954
Fixes https://github.com/terrycain/aioboto3/issues/265

### Assumptions
That if `"Transfer-Encoding": "Chunked"` is always removed from the header dict before being passed to aiohttp is ok. (aiohttp reallllllly doesnt like it if you send it streaming body, and already set the transfer encoding :/ ). This ends up being injected from multiple places, mainly some of the checksum code as wells somewhere deep in the aws prepare_request() functions,

---

AWS Page - https://aws.amazon.com/blogs/aws/new-additional-checksum-algorithms-for-amazon-s3/

### CODE REVIEW

**Comment by terrycain** [comment]
@thehesiod am not too pleased with `headers.pop('Transfer-Encoding', '')` in `httpsession.py` you think theres a better way to do it?

**Comment by terrycain** [comment]
So when I run the S3 tests locally without the changes, I get the same number failed as with the changes, so I don't think they're related :)

**Comment by terrycain** [comment]
Yeah it looks like moto is unhappy for some reason. I ran the server standalone and pointed the tests at it and they seem to be getting 404'd at least on creating multipart uploads, even using boto3.

**Comment by thehesiod** [comment]
thanks for all the research Terry! hopefully get to this soon, as it is getting that last release out it's 3am and I need to wake up at 7am, yowzers ;)

**Comment by thehesiod** [comment]
looks like they pinned werkzeug + flask and I bumped us to the latest moto, may want to try with what we have and rebase to master

**Comment by thehesiod** [comment]
just read a bit about this feature, very cool!  This will solve the whole async file streaming support, since otherwise it didn't make much sense.  I'm very interested in getting this working asap so will put further priority on reviewing this

**Comment by thehesiod** [comment]
I could have used this feature several times in the last few years :)

**Comment by terrycain** [comment]
Yeah the coverage isnt going to be great as we can't test this unless moto runs with TLS. We could make that would though I think that should be done separately.

**Comment by thehesiod** [comment]
hey @terrycain for the delay, been nuts. Hey some good news I got my first sponsor for this project, can I share a few bucks with you?

**Comment by terrycain** [comment]
Sure :D always up for some free money [this](https://www.paypal.me/terrycain1337) is my paypal link else I've finally had the will to setup the github sponsors stuff :).

**Comment by coolacid** [comment]
I dislike being "that guy" but wanted to check in to see what needs to happen to get this in? Looking forward to this fix.

**Comment by terrycain** [comment]
@thehesiod can we get this merged ignoring codecov?

**Comment by thehesiod** [comment]
hey sorry been nuts, got covid and perhaps RSV too from our kids one of my highest priorities is getting this in, really want to look at this in detail

**Comment by thehesiod** [comment]
ok changes look good, just want to dig into the Transfer-Encoding issue

**Comment by thehesiod** [comment]
ok figured out the chunk thing, and added support for running https tests, unfortunately moto doesn't appear to support the algs yet so still need to xfail for now.

File: aiobotocore/httpsession.py [hunk 1]
```diff
@@ -188,20 +188,27 @@ async def send(self, request):
                 host = urlparse(request.url).hostname
                 proxy_headers['host'] = host
 
-            # https://github.com/boto/botocore/issues/1255
-            headers['Accept-Encoding'] = 'identity'
```
Comment (thehesiod) [inline aiobotocore/httpsession.py:192]: don't modify param passed in, so moved down

File: aiobotocore/httpsession.py [hunk 2]
```diff
@@ -188,20 +188,27 @@ async def send(self, request):
...
-
-            headers_ = MultiDict(
+            headers_ = CIMultiDict(
```
Comment (thehesiod) [inline aiobotocore/httpsession.py:191]: incoming headers is case insensitive, so copy should be as well

File: aiobotocore/httpsession.py [hunk 3]
```diff
@@ -188,20 +188,27 @@ async def send(self, request):
...
             )
 
+            # https://github.com/boto/botocore/issues/1255
+            headers_['Accept-Encoding'] = 'identity'
+
+            chunked = None
+            if headers_.get('Transfer-Encoding', '').lower() == 'chunked':
+                # aiohttp wants chunking as a param, and not a header
+                headers_.pop('Transfer-Encoding', '')
+                chunked = True
```
Comment (thehesiod) [inline aiobotocore/httpsession.py:202]: aiohttp wants chunked to be passed as a param, not a header, so upgrade it

**Comment by terrycain** [comment]
:tada:

**Comment by coolacid** [comment]
Confirmed this fixed setting checksum algorithm when using S3FS. Thanks!

**Pull request merged**